### PR TITLE
RFE: add test support for session ID user filter

### PR DESCRIPTION
**Pull request by rgbriggs**
```
test: RFE: add a session ID filter to the kernel's user filter
https://github.com/linux-audit/audit-kernel/issues/4
```

Signed-off-by: Richard Guy Briggs rgb@redhat.com

### CODE REVIEW

File: tests/sessionid_filter/test [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+$result = system("ausearch -i -m SYSCALL -sc open -p $pid --session $sessionid -k $key > $stdout 2> $stderr");
+ok($result, 0);
+
+# test if we generate the PATH record correctly
+my $line;
+my $found_msg = 0;
+while ($line = <$fh_out>) {
+	# test if we generate a PATH record
+	if ($line =~ m?^type=PATH ? and
+	    $line =~ m? name=/tmp/$key ? and
```
Comment (pcmoore) [inline tests/sessionid_filter/test]: In addition to matching on the key we should also match on session ID given the nature of the test ... and yes, I realize you are passing the session ID to ausearch already.

**Comment by rgbriggs** [comment]
Since the key is only in the SYSCALL message type and not in the PATH, CWD or PROCTITLE message types, I'd have to match on two messages from the same record, which is doable, but more complex than I figured was necessary since we are already requesting only records with a matching session ID, PID and key.  How will this modification improve on what is already coded?  Would adding "-ts recent" to the ausearch command be appropriate?

**Comment by pcmoore** [comment]
> Since the key is only in the SYSCALL message type and not in the PATH, CWD or PROCTITLE
> message types, I'd have to match on two messages from the same record, which is doable, but more 
> complex than I figured was necessary ....

Simply match on the SYSCALL record, this particular test does not need to match on the pathname; I would say that they key and session ID are the important fields for this test.

> How will this modification improve on what is already coded?

It would ensure a positive match on the session ID, which is the point of this test.

> Would adding "-ts recent" to the ausearch command be appropriate?

I don't think it matters, I would assume leave it out as passing the key filter to ausearch should help limit the results.

**Comment by rgbriggs** [comment]
Ok, pushed an update that adds an explicit check of the session ID in the SYSCALL message.

**Comment by pcmoore** [comment]
Wait, why are we still testing for a PATH record?  A SYSCALL record with the correct session ID should be sufficient, yes?

**Comment by pcmoore** [comment]
@rgbriggs See above.  Why do we care about the PATH record in this test?

**Comment by rgbriggs** [comment]
Ok, strictly speaking that PATH record test doesn't belong in this one, but should be moved/retained for a dedicated test...

**Comment by pcmoore** [comment]
We already have several tests that check for the presence and correctness of PATH records, see the file_* tests.

**Comment by rgbriggs** [comment]
There's a new version of the test script available here: https://github.com/rgbriggs/audit-testsuite/tree/ghak4-test-for-sessionID-user-filter
This doesn't yet address SESSIONID_SET.

**Comment by rgbriggs** [comment]
There's a new version of the test script that now addresses SESSIONID_SET.

**Comment by rgbriggs** [comment]
And yet another newer version of this test script at https://github.com/rgbriggs/audit-testsuite/commit/c265e9fc5a7cd25315baec67f36d29ae0b11c020

It isn't tracking properly because I renamed the branch...  I suppose I could push it to the lagh4... branch.

**Comment by pcmoore** [comment]
No worries on the tracking, links to the mailing list posts are probably more helpful anyway as they capture the discussion.

File: tests/sessionid_filter/Makefile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+TARGETS=$(patsubst %.c,%,$(wildcard *.c))
+
+LDLIBS += -lpthread
+
+all: $(TARGETS)
+clean:
+	rm -f $(TARGETS)
+
```
Comment (pcmoore) [inline tests/sessionid_filter/Makefile]: Let's rename this directory to "filter_sessionid" (see my comments on the exclude filter patch for the testsuite).  Also, don't forget to wire this into the top level Makefile so it gets executed.

File: tests/sessionid_filter/test [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+# test for the SYSCALL message
+$result = system("ausearch -i -m SYSCALL -sc open -p $pid --session $sessionid -k $key > $stdout 2> $stderr");
+ok($result, 0);
+
+# test if we generate the PATH record correctly
+my $line;
+my $found_path_msg = 0;
+my $syscall_msg_match = 0;
+while ($line = <$fh_out>) {
+	# test if we generate a PATH record
```
Comment (pcmoore) [inline tests/sessionid_filter/test]: We don't really need to check for PATH records here, we have a separate set of tests for that.  Check for the SYSCALL record and call it good; we don't want a problem with PATH records to also cause a test failure here.

**Comment by pcmoore** [comment]
Commit https://github.com/rgbriggs/audit-testsuite/commit/13f5d6f31ffb542cbdc3523716e3640a7d71a344 looks reasonable, although I would prefer if we renamed the directory from "sessionid_filter" to "filter_sessionid".  Does that sound reasonable @rgbriggs ?

**Comment by pcmoore** [comment]
I added one comment to the https://github.com/rgbriggs/audit-testsuite/commit/c265e9fc5a7cd25315baec67f36d29ae0b11c020 patch.

**Comment by rgbriggs** [comment]
I've already argued against flipping the name around, but I don't care enough to press it...  This pull request is stale since my branch got renamed on my local machine to reflect the new naming convention.  The up to date branch is: https://github.com/rgbriggs/audit-testsuite/tree/ghak4-test-for-sessionID-user-filter

**Comment by rgbriggs** [comment]
On 2016-08-30 12:41, Paul Moore wrote:
> I added one comment to the https://github.com/rgbriggs/audit-testsuite/commit/c265e9fc5a7cd25315baec67f36d29ae0b11c020 patch.

This is also stale.


- RGB

--
Richard Guy Briggs <rgb@redhat.com>
Kernel Security Engineering, Base Operating Systems, Red Hat
Remote, Ottawa, Canada
Voice: +1.647.777.2635, Internal: (81) 32635

**Comment by pcmoore** [comment]
For reference, the latest draft from @rgbriggs appears to be here: https://github.com/rgbriggs/audit-testsuite/commit/2584bcc842da7511dbac3482f2eaf2afc6f2db39

**Comment by pcmoore** [comment]
I made one small comment on this revision, once we resolve this I think we are good to go as soon as the kernel/userspace patches are merged.

**Comment by rgbriggs** [comment]
@pcmoore ok grab https://github.com/rgbriggs/audit-testsuite/commit/ec6d30f873f43d2aa83c64c3384315ab86eafaf5

**Comment by pcmoore** [comment]
I'll trust that the perl expression matching is correct.  Looks like we are just waiting on Steve now.

**Comment by pcmoore** [comment]
Adjusting the pending label, it looks like we are just waiting on a rework for the tests to remove the set flag.  Also, a reminder to @rgbriggs that we need this test and RFE wiki page done before I send the kernel patches to Linus.

**Comment by rgbriggs** [comment]
Both the test update and push and the RFE update were done last week.

**Comment by rgbriggs** [comment]
https://github.com/rgbriggs/audit-testsuite/commit/94e418b8812a091d30c4cd6ace86f4378a85ddb1

**Comment by pcmoore** [comment]
@rgbriggs great, I'll take a look at them today.  For future reference, when you update a PR I think it would be good practice to comment on the PR/thread as GitHub doesn't appear to send notifications that you updated the code in your PR.

**Comment by pcmoore** [comment]
The test looks good to me, but since we don't have a released version of the audit userspace that supports the session ID filter (waiting on audit userspace v2.7) I'm not going to merge this just yet as I don't want to keep seeing failures when I run the testsuite.

@stevegrubb has mentioned that he intends to release the audit userspace v2.7 this month (December 2016) so we shouldn't have to wait long.

**Comment by pcmoore** [comment]
Merged via b3782254c52f1f63a6a42e8e821e60a5d62b88a9.

**Pull request closed without merging**